### PR TITLE
Use `getCmp`

### PR DESCRIPTION
**Pull request by oliverlloyd**
## What does this change?
Implement the `getCmp` function as a replacement for importing the `cmp` object each time.

## Why?
We don't want to repeatedly import the cmp code from the npm lib and there's already a copy of it set on the window object for us to access so we're leaning into this pattern.

### CODE REVIEW

File: dotcom-rendering/src/web/components/SignInGate/displayRule.ts [hunk 1]
```diff
@@ -3,10 +3,10 @@ import {
 	DailyArticle,
 	getDailyArticleCount,
 } from '@frontend/web/lib/dailyArticleCount';
-import { onConsentChange } from '@guardian/consent-management-platform';
 import { ConsentState } from '@guardian/consent-management-platform/dist/types';
 import { getLocale } from '@guardian/libs';
 import { hasUserDismissedGateMoreThanCount } from '@root/src/web/components/SignInGate/dismissGate';
+import { guCmp } from '../../lib/guCmp';
```
[the remark this answers is not in the document]
Comment (oliverlloyd) [reply dotcom-rendering/src/web/components/SignInGate/displayRule.ts:9]: We recently [removed these aliases from DCR](https://github.com/guardian/dotcom-rendering/pull/3936)

[the remark this answers is not in the document]
Comment (tjmw) [reply dotcom-rendering/src/web/components/SignInGate/displayRule.ts:9]: Ahh missed that 👍

File: dotcom-rendering/src/web/lib/braze/hasRequiredConsents.test.ts [hunk 2]
```diff
@@ -1,94 +0,0 @@
-import { hasRequiredConsents } from './hasRequiredConsents';
```
[the remark this answers is not in the document]
Comment (oliverlloyd) [reply dotcom-rendering/src/web/lib/braze/hasRequiredConsents.test.ts:1]: Yes, please. I actually have this task as a todo on this PR (but only in my head/some text document on my laptop - I could probably surface this more!)

I was hoping that there would be a way to implement a Cypress test that verified the UI responded as expected based on different consent settings.

File: dotcom-rendering/src/web/lib/guCmp.ts [hunk 3]
```diff
@@ -22,3 +33,5 @@ export const getCmp = () => {
 	}
 	return window.guCmpHotFix;
 };
+
+export const guCmp = getCmp();
```
[the remark this answers is not in the document]
Comment (oliverlloyd) [reply dotcom-rendering/src/web/lib/guCmp.ts:37]: lol, not sure how I ended up with that pattern! Fixed now.

[the remark this answers is not in the document]
Comment (oliverlloyd) [reply dotcom-rendering/src/web/lib/guCmp.ts]: In the end I rationalised the approach here to something a bit more semantic; I'm now using `getCmp` and executing it in the files each time.

**Comment by sndrs** [comment]
this feels very brittle to me – `guCmpHotFix` is a temporary, internal, undocumented and unsupported implementation detail of the CMP so it can handle being unexpectedly instantiated twice, and liable to disappear without warning

it is 100% not intended to be the public interface (unless this has changed @shtukas @kenoir?)

> We don't want to repeatedly import the cmp code from the npm lib 

why is this? there should always be only one instance of the code available to DCR? what's the overhead?

> there's already a copy of it set on the window object for us to access so we're leaning into this pattern.

as mentioned above, there _is_ but it's undocumented and not intended for external use

**Comment by oliverlloyd** [comment]
> what's the overhead?

Fair question. And I'm actually super happy for this approach to be challenged in general.

The motivation for this work was rooted in the cmp lib not being server safe so when it's an import that gets evaluated and throws errors whenever we server side render components. We've been getting away with this up to now by using Portals and dynamic imports but with the transition to islands we started to hit this problem much more often (the island version of a Portal evaluates the component on the server).

So, having this window prop solves this in a nice, low effort way but maybe the larger solution is to make cmp server safe?

**Comment by ashishpuliyel** [comment]
I'm not sure how meaningful it is to have a 'server safe' cmp (since all the meaningful state is clientside). However I wonder if this should be internal to the CMP. As in consumers shouldn't have to worry about getting the instance, or making sure it's the instance it should take care of it itself? 

I echo Alex's concern about increasingly relying on something called 'hotfix' but I think the solution is to rename it, formally make the CMP library use it internally, and have it check itself internally on _any_ method call, transparent to rh consumer. And the consumer just calls things blithely, safe in the knowledge that the CMP internally ensures it's refering to a single instance of its state.  

I can do a quick mockup of what I'm talking about of I'm not being clear here (my code is sometimes more articulate than I am).

**Comment by oliverlloyd** [comment]
Closing now that we have #4126 🎉

**Pull request closed without merging**